controller/identity/validator.go

```diff
@@ -0,0 +1,97 @@
+package identity
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	kauthnApi "k8s.io/api/authentication/v1"
+	kauthzApi "k8s.io/api/authorization/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	kauthn "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	kauthz "k8s.io/client-go/kubernetes/typed/authorization/v1"
+
+	"github.com/linkerd/linkerd2/pkg/identity"
+)
+
+// K8sTokenValidator implements Validator for Kubernetes bearer tokens.
+type K8sTokenValidator struct {
+	authn  kauthn.AuthenticationV1Interface
+	domain *TrustDomain
+}
+
+// NewK8sTokenValidator takes a kubernetes client and trust domain to create a
+// K8sTokenValidator.
+//
+// The kubernetes client is used immediately to validate that the client has
+// sufficient privileges to perform token reviews. An error is returned if this
+// access check fails.
+func NewK8sTokenValidator(
+	k8s k8s.Interface,
+	domain *TrustDomain,
+) (*K8sTokenValidator, error) {
+	if err := checkAccess(k8s.AuthorizationV1()); err != nil {
+		return nil, err
+	}
+
+	authn := k8s.AuthenticationV1()
+	return &K8sTokenValidator{authn, domain}, nil
+}
+
+func init() {
+	var _ identity.Validator = &K8sTokenValidator{}
```

Comment:
I can't figure out what this `init()` function is doing. Mind adding a comment to explain it?

Reply:
it's asserting that the type implements the interface

Reply:
Ah, thanks for the explanation. It would be more conventional to have the return type of `NewK8sTokenValidator` be `identity.Validator`, which would ensure that `K8sTokenValidator` satisfies the interface. I'm thinking something like:

```diff
diff --git a/controller/identity/validator.go b/controller/identity/validator.go
index 76554444..e6115ad3 100644
--- a/controller/identity/validator.go
+++ b/controller/identity/validator.go
@@ -29,7 +29,7 @@ type K8sTokenValidator struct {
 func NewK8sTokenValidator(
 	k8s k8s.Interface,
 	domain *TrustDomain,
-) (*K8sTokenValidator, error) {
+) (identity.Validator, error) {
 	if err := checkAccess(k8s.AuthorizationV1()); err != nil {
 		return nil, err
 	}
@@ -38,11 +38,6 @@ func NewK8sTokenValidator(
 	return &K8sTokenValidator{authn, domain}, nil
 }
 
-func init() {
-	// Assert that the struct implements the interface.
-	var _ identity.Validator = &K8sTokenValidator{}
-}
-
 // Validate accepts kubernetes bearer tokens and returns a DNS-form linkerd ID.
 func (k *K8sTokenValidator) Validate(_ context.Context, tok []byte) (string, error) {
 	// TODO: Set/check `audience`
```